File: client/src/services/review.service.ts
import { apiService } from './api';

export interface Review {
  reviewId: number;
  rating: number;
  comment: string;
  createdAt: string;
  lastUpdated: string;
  userId: number;
  userName: string;
  productId: number;
  productName: string;
}

export interface ReviewRequest {
  productId: number;
  rating: number;
  comment?: string;
}

export const reviewService = {
  /**
   * Tạo review mới
   */
  createReview: async (request: ReviewRequest): Promise<Review> => {
    return apiService.post<Review>('/reviews', request);
  },

  /**
   * Lấy tất cả review của một sản phẩm
   */
  getReviewsByProduct: async (productId: number): Promise<Review[]> => {
    return apiService.get<Review[]>(`/reviews/product/${productId}`);
  },

  /**
   * Lấy tất cả review của một user
   */
  getReviewsByUser: async (userId: number): Promise<Review[]> => {
    return apiService.get<Review[]>(`/reviews/user/${userId}`);
  },

  /**
   * Kiểm tra user đã đánh giá sản phẩm chưa
   */
  hasUserReviewedProduct: async (productId: number): Promise<boolean> => {
    return apiService.get<boolean>(`/reviews/check/${productId}`);
  },
};

File: server/src/main/java/iuh/fit/server/services/ReviewService.java
package iuh.fit.server.services;

import iuh.fit.server.dto.request.ReviewRequest;
import iuh.fit.server.dto.response.ReviewResponse;

import java.util.List;

public interface ReviewService {
    /**
     * Tạo review mới
     * @param userId ID của user
     * @param request Review request
     * @return Review response
     */
    ReviewResponse createReview(Integer userId, ReviewRequest request);
    
    /**
     * Lấy tất cả review của một sản phẩm
     * @param productId ID của sản phẩm
     * @return Danh sách review
     */
    List<ReviewResponse> getReviewsByProduct(Integer productId);
    
    /**
     * Lấy tất cả review của một user
     * @param userId ID của user
     * @return Danh sách review
     */
    List<ReviewResponse> getReviewsByUser(Integer userId);
    
    /**
     * Kiểm tra user đã đánh giá sản phẩm chưa
     * @param userId ID của user
     * @param productId ID của sản phẩm
     * @return true nếu đã đánh giá
     */
    boolean hasUserReviewedProduct(Integer userId, Integer productId);
    
    /**
     * Kiểm tra user đã mua sản phẩm chưa (có order với status PAID)
     * @param userId ID của user
     * @param productId ID của sản phẩm
     * @return true nếu đã mua
     */
    boolean hasUserPurchasedProduct(Integer userId, Integer productId);
}

